pkg/cluster/internal/create/kubeadm-config.go

```diff
@@ -68,6 +68,12 @@ func runKubeadmConfig(ec *execContext, configNode *NodeReplica) error {
 		return errors.Wrap(err, "failed to get kubernetes version from node: %v")
 	}
 
+	// get the address the API Server will advertise to other members of the cluster
+	apiAdvertiseAddress, err := node.IP(ec.Context.IPv6)
+	if err != nil {
+		return errors.Wrap(err, "failed to get IP for bootsrap node")
+	}
+
 	// get the control plane endpoint, in case the cluster has an external load balancer in
 	// front of the control-plane nodes
 	controlPlaneEndpoint, err := getControlPlaneEndpoint(ec)
```

Comment:
`getControlPlaneEndpoint` might need modifications based on IPv6.

have you tested an HA setup?

Reply:
didn't test it, but will do.
I've already modified the function `func getControlPlaneEndpoint(` to support IPv6, check L135

Reply:
@ncdc results of an HA setup test

```
 # docker ps
CONTAINER ID        IMAGE                  COMMAND                  CREATED             STATUS              PORTS                                NAMES
7b37bcbcf3b6        kindest/node:v1.13.3   "/usr/local/bin/entr…"   11 minutes ago      Up 11 minutes                                            kind-worker2
89fa42fd404e        kindest/node:v1.13.3   "/usr/local/bin/entr…"   12 minutes ago      Up 12 minutes                                            kind-worker1
0d0caea990ca        kindest/node:v1.13.3   "/usr/local/bin/entr…"   12 minutes ago      Up 12 minutes       33453/tcp, 0.0.0.0:33453->6443/tcp   kind-control-plane3
422bd7c52d53        kindest/node:v1.13.3   "/usr/local/bin/entr…"   13 minutes ago      Up 12 minutes       34151/tcp, 0.0.0.0:34151->6443/tcp   kind-control-plane2
780f8cde9d6d        kindest/node:v1.13.3   "/usr/local/bin/entr…"   13 minutes ago      Up 13 minutes       37927/tcp, 0.0.0.0:37927->6443/tcp   kind-control-plane1
3f2e867b33b6        kindest/node:v1.13.3   "/usr/local/bin/entr…"   13 minutes ago      Up 13 minutes       40189/tcp, 0.0.0.0:40189->6443/tcp   kind-lb
```
```
# kubectl get endpoints --all-namespaces
NAMESPACE     NAME                      ENDPOINTS                                                                                   AGE
default       kubernetes                [2001:db8:1::242:ac11:3]:6443,[2001:db8:1::242:ac11:4]:6443,[2001:db8:1::242:ac11:5]:6443   11m
kube-system   kube-controller-manager   <none>                                                                                      10m
kube-system   kube-dns                  10.32.0.2:53,10.32.0.3:53,10.32.0.2:53 + 1 more...                                          10m
kube-system   kube-scheduler            <none>                                                                                      10m
```

with current parameters:
`./kind create cluster --ipv6 --config config-ipv6.yaml`

and config-ipv6.yaml

```
# this config file contains all config fields with comments
kind: Config
apiVersion: kind.sigs.k8s.io/v1alpha2
# 3 control plane node and 3 workers
nodes:
# the control plane node config
- role: control-plane
  replicas: 3
  # patch the generated kubeadm config with some extra settings
  kubeadmConfigPatches:
  - |
    apiVersion: kubeadm.k8s.io/v1beta1
    kind: ClusterConfiguration
    metadata:
      name: config
    networking:
      serviceSubnet: "fd00:1234::/112"
# the three workers
- role: worker
  # replicas specifes the number of nodes to create with this configuration
  replicas: 2
- role: external-load-balancer
```

@BenTheElder I'd like to add some e2e scenarios, could you help me or guide me?
Should those tests be part of this PR?

Reply:
@aojea 
hm, given weave supposedly does not support ipv6, how are you testing this?

> ./kind create cluster --ipv6

we need to remove the `--ipv6` flag from the CLI, should be config only.
the CLI is only for the basic functionality.

> @BenTheElder I'd like to add some e2e scenarios, could you help me or guide me?
Should those tests be part of this PR?

let's get the support in kind first, and then we can think about e2e.
ideally things like ipv6 should be owned by sig-network, but i can think we might be able to add this in the sig-cluster-lifecycle dashboard too.

i will try your PR later today myself, hopefully.

Reply:
@neolit123 seems that the cluster control plane is ipv6 abut the data plane is ipv4, let me do more tests changing to the cni bridge plugin and come back to you

Reply:
@neolit123 I can access IPv6 services with the new [CNI plugin I've created](https://github.com/aojea/kindnet) :)

To test it just update the `node-image` with the new `kind` binary built from this PR
`kind build node-image --base-image kindest/base:latest --type bazel`
And spawn a cluster with the following config file 
```
# this config file contains all config fields with comments
kind: Config
apiVersion: kind.sigs.k8s.io/v1alpha2
# 3 control plane node and 3 workers
nodes:
# the control plane node config
- role: control-plane
  replicas: 1
  # patch the generated kubeadm config with some extra settings
  kubeadmConfigPatches:
  - |
    apiVersion: kubeadm.k8s.io/v1beta1
    kind: ClusterConfiguration
    metadata:
      name: config
    networking:
      podSubnet: "fd00:100::/64"
      serviceSubnet: "fd00:1234::/112"
# the three workers
- role: worker
  # replicas specifes the number of nodes to create with this configuration
  replicas: 2
```

`kind create cluster --ipv6 --config config-ipv6-ci.yaml --loglevel debug --image kindest/node:latest`

and you can have an IPv6 k8s cluster:

```
 kubectl get nodes -o wide
NAME                 STATUS    ROLES     AGE       VERSION          INTERNAL-IP              EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION                CONTAINER-RUNTIME
kind-control-plane   Ready     master    3m9s      v1.14.0-beta.1   2001:db8:1::242:ac11:3   <none>        Ubuntu 18.04.1 LTS   4.12.14-lp150.12.45-default   docker://18.6.3
kind-worker1         Ready     <none>    2m31s     v1.14.0-beta.1   2001:db8:1::242:ac11:4   <none>        Ubuntu 18.04.1 LTS   4.12.14-lp150.12.45-default   docker://18.6.3
kind-worker2         Ready     <none>    2m16s     v1.14.0-beta.1   2001:db8:1::242:ac11:5   <none>        Ubuntu 18.04.1 LTS   4.12.14-lp150.12.45-default   docker://18.6.3
```
with pod IPv6 connectivity and services
```
kubectl -n kube-system get ep -o wide
NAME                      ENDPOINTS                                                        AGE
kube-controller-manager   <none>                                                           4m34s
kube-dns                  [fd00:100::2]:53,[fd00:100::3]:53,[fd00:100::2]:53 + 3 more...   4m17s
kube-scheduler            <none>                                                           4m36s
```
The pods obtain IPv6 addresses:

```
 kubectl run -i --tty busybox12 --image=busybox --restart=Never -- sh
If you don't see a command prompt, try pressing enter.
/ # ip -6 a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qlen 1000
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
5: eth0@if6: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500
    inet6 fd00:100::1:0:0:2/80 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::5c62:8aff:fe75:7ce0/64 scope link
       valid_lft forever preferred_lft forever
```
with pod connectivity between nodes

```
/ # ping fd00:100::2
PING fd00:100::2 (fd00:100::2): 56 data bytes
64 bytes from fd00:100::2: seq=0 ttl=62 time=0.181 ms
64 bytes from fd00:100::2: seq=1 ttl=62 time=0.151 ms
```

and pod to service connectivity
```
/ # nc -zv fd00:1234::a 53
fd00:1234::a ([fd00:1234::a]:53) open
```

Reply:
@aojea nice.

>       podSubnet: "fd00:100::/64"
>      serviceSubnet: "fd00:1234::/112"

do we have to instruct the users to patch the config with IPv6 CIDRs if they want IPv6 support.

> kind create cluster --ipv6 

we might want to move this a flag in the kind Config.

Reply:
Where should be the best place to add the ipv6 flag?
What do you think to add it in the new `networking` section of the v1alpha3 API https://github.com/kubernetes-sigs/kind/issues/340 ?

Reply:
> What do you think to add it in the new networking section of the v1alpha3 API

yes, makes sense.

and how about this?
> do we have to instruct the users to patch the config with IPv6 CIDRs if they want IPv6 support?

Reply:
@neolit123 that's a good question but I miss context here, but I think that maybe can be a topic to be discussed in `kubeadm`.
I could check that `kubeadm` sets a default `servicesubnet` if none is provided, should we do the same for IPv6? if that's the case we should add a flag to `kubeadm` then

Reply:
> I could check that kubeadm sets a default servicesubnet if none is provided, should we do the same for IPv6? if that's the case we should add a flag to kubeadm then

here are kubeadm defaults for these CIDRs:
https://github.com/kubernetes/kubernetes/blob/b66e332d3c19ce78e00dd7c904fa29e8f6784ba0/cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go#L31-L34

it not very likely that we will add a flag or a field in the kubeadm config for this.
IPv6 is explicit on the kubeadm side and up to the administrator to decide what stack to use. so kind as a kubeadm wrapper might have to handle this explicitly (as your patch is doing here).

at least that's how i see it for now.